--- a/node-yarn/lib/cli.js
+++ b/node-yarn/lib/cli.js
@@ -36003,7 +36003,7 @@
 var getBinFolder = exports.getBinFolder = function () {
   var _ref8 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)(function* (config, flags) {
     var prefix = yield getGlobalPrefix(config, flags);
-    return path.resolve(prefix, 'bin');
+    return path.resolve('/usr/local/bin');
   });
 
   return function getBinFolder(_x5, _x6) {
