--- a/configure.ac
+++ b/configure.ac
@@ -1441,7 +1441,7 @@ if test "$Py_LTO" = 'true' ; then
           LTOCFLAGS="-flto"
           ;;
         *)
-          LTOFLAGS="-flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none"
+          LTOFLAGS="-g -O3 -fwrapv -flto=1 -fuse-linker-plugin -ffat-lto-objects -flto-partition=none"
           ;;
       esac
       ;;
@@ -1519,8 +1519,8 @@ case $CC in
         fi
         ;;
       *)
-        PGO_PROF_GEN_FLAG="-fprofile-generate"
-        PGO_PROF_USE_FLAG="-fprofile-use -fprofile-correction"
+        PGO_PROF_GEN_FLAG="-fprofile-generate -fprofile-update=prefer-atomic"
+        PGO_PROF_USE_FLAG="-fprofile-use -fprofile-correction -fprofile-partial-training -Wmissing-profile -Wno-error=coverage-mismatch"
         LLVM_PROF_MERGER="true"
         LLVM_PROF_FILE=""
         ;;
