--- a/Makefile.pre.in
+++ b/Makefile.pre.in
@@ -522,6 +522,14 @@ build_all_generate_profile:
 
 run_profile_task:
 	@ # FIXME: can't run for a cross build
+	if [ '$(PGO_EXTRA_PASS)' = 1 ] ; then \
+	  ./pyperformance.sh venv create --python "$${PWD}/$(BUILDPYTHON)" ; \
+	  for i in 1 2 ; do \
+	    ./pyperformance.sh run --debug-single-value --python "$${PWD}/$(BUILDPYTHON)" || true ; \
+	    unset CONTAINER_PYTHON_COMPAT ; \
+	  done ; \
+	  export CONTAINER_PYTHON_COMPAT=1 ; \
+	fi ; \
 	$(LLVM_PROF_FILE) $(RUNSHARED) ./$(BUILDPYTHON) $(PROFILE_TASK) || true
 
 build_all_merge_profile:
--- a/configure.ac
+++ b/configure.ac
@@ -1344,7 +1344,7 @@ AC_ARG_VAR(PROFILE_TASK, Python args for
 AC_MSG_CHECKING(PROFILE_TASK)
 if test -z "$PROFILE_TASK"
 then
-	PROFILE_TASK='-m test --pgo'
+	PROFILE_TASK='-m test --pgo-extended'
 fi
 AC_MSG_RESULT($PROFILE_TASK)
 
