--- a/lib/internal/modules/cjs/loader.js
+++ b/lib/internal/modules/cjs/loader.js
@@ -1119,9 +1119,9 @@ Module._initPaths = function() {
     postDirs.push(path.join('lib', 'node'));
   }
 
-  let paths = postDirs.map(postDir => {
-    return path.resolve(prefixDir, postDir);
-  });
+  let paths = [path.resolve(prefixDir, 'lib', 'node_modules')];
+  // compatibility with Debian
+  paths = paths.concat(postDirs.map(postDir => { return path.resolve('/usr', postDir); }));
 
   if (homeDir) {
     paths.unshift(path.resolve(homeDir, '.node_libraries'));
