--- a/pip-numpy/benchmarks/benchmarks/bench_linalg.py
+++ b/pip-numpy/benchmarks/benchmarks/bench_linalg.py
@@ -102,7 +102,7 @@ class LinalgSmallArrays(Benchmark):
     """ Test overhead of linalg methods for small arrays """
     def setup(self):
         self.array_5 = np.arange(5.)
-        self.array_5_5 = np.arange(5.)
+        self.array_5_5 = self.array_5 + self.array_5[:, None]
 
     def time_norm_small_array(self):
         np.linalg.norm(self.array_5)
