--- a/configure.ac
+++ b/configure.ac
@@ -154,9 +154,6 @@ AC_ARG_WITH([build-python],
     AS_VAR_IF([with_build_python], [yes], [with_build_python=python$PACKAGE_VERSION])
     AS_VAR_IF([with_build_python], [no], [AC_MSG_ERROR([invalid --with-build-python option: expected path or "yes", not "no"])])
 
-    if ! $(command -v "$with_build_python" >/dev/null 2>&1); then
-      AC_MSG_ERROR([invalid or missing build python binary "$with_build_python"])
-    fi
     build_python_ver=$($with_build_python -c "import sys; print(f'{sys.version_info.major}.{sys.version_info.minor}')")
     if test "$build_python_ver" != "$PACKAGE_VERSION"; then
       AC_MSG_ERROR(["$with_build_python" has incompatible version $build_python_ver (expected: $PACKAGE_VERSION)])
