--- a/Python/initconfig.c
+++ b/Python/initconfig.c
@@ -1669,7 +1669,7 @@ config_read_env_vars(PyConfig *config)
     _Py_get_env_flag(use_env, &config->optimization_level, "PYTHONOPTIMIZE");
     _Py_get_env_flag(use_env, &config->inspect, "PYTHONINSPECT");
 
-    int dont_write_bytecode = 0;
+    int dont_write_bytecode = 1;
     _Py_get_env_flag(use_env, &dont_write_bytecode, "PYTHONDONTWRITEBYTECODE");
     if (dont_write_bytecode) {
         config->write_bytecode = 0;
@@ -1681,7 +1681,7 @@ config_read_env_vars(PyConfig *config)
         config->user_site_directory = 0;
     }
 
-    int unbuffered_stdio = 0;
+    int unbuffered_stdio = 1;
     _Py_get_env_flag(use_env, &unbuffered_stdio, "PYTHONUNBUFFERED");
     if (unbuffered_stdio) {
         config->buffered_stdio = 0;
