--- a/Makefile.pre.in
+++ b/Makefile.pre.in
@@ -1912,13 +1912,10 @@ clean-retain-profile: pycremoval
 	find . -name '*.s[ol]' -exec rm -f {} ';'
 	find . -name '*.so.[0-9]*.[0-9]*' -exec rm -f {} ';'
 	find . -name '*.lst' -exec rm -f {} ';'
-	find build -name 'fficonfig.h' -exec rm -f {} ';' || true
-	find build -name '*.py' -exec rm -f {} ';' || true
-	find build -name '*.py[co]' -exec rm -f {} ';' || true
+	find . -name '*.py[co]' -exec rm -f {} ';' || true
 	-rm -f pybuilddir.txt
 	-rm -f Lib/lib2to3/*Grammar*.pickle
 	-rm -f Programs/_testembed Programs/_freeze_importlib
-	-find build -type f -a ! -name '*.gc??' -exec rm -f {} ';'
 	-rm -f Include/pydtrace_probes.h
 	-rm -f profile-gen-stamp
 
