--- a/configure.ac
+++ b/configure.ac
@@ -1438,7 +1438,7 @@ if test "$Py_LTO" = 'true' ; then
           LTOCFLAGS="-flto"
           ;;
         *)
-          LTOFLAGS="-flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none"
+          LTOFLAGS="-g -O3 -fwrapv -flto=2 -fuse-linker-plugin -ffat-lto-objects -flto-partition=none"
           ;;
       esac
       ;;
@@ -1516,8 +1516,8 @@ case $CC in
         fi
         ;;
       *)
-        PGO_PROF_GEN_FLAG="-fprofile-generate"
-        PGO_PROF_USE_FLAG="-fprofile-use -fprofile-correction"
+        PGO_PROF_GEN_FLAG="-fprofile-generate -fprofile-update=prefer-atomic"
+        PGO_PROF_USE_FLAG="-fprofile-use -fprofile-correction -fprofile-partial-training -Wmissing-profile -Wno-error=coverage-mismatch"
         LLVM_PROF_MERGER="true"
         LLVM_PROF_FILE=""
         ;;
--- a/Makefile.pre.in
+++ b/Makefile.pre.in
@@ -722,7 +722,7 @@ Makefile Modules/config.c: Makefile.pre
 
 
 Programs/_testembed: Programs/_testembed.o $(LIBRARY_DEPS)
-	$(LINKCC) $(PY_CORE_LDFLAGS) $(LINKFORSHARED) -o $@ Programs/_testembed.o $(BLDLIBRARY) $(LIBS) $(MODLIBS) $(SYSLIBS)
+	$(LINKCC) -fno-lto $(filter-out -flto% -ffat-lto-objects,$(PY_CORE_LDFLAGS) $(LINKFORSHARED)) -o $@ Programs/_testembed.o $(BLDLIBRARY) $(LIBS) $(MODLIBS) $(SYSLIBS)
 
 ############################################################################
 # Importlib
@@ -730,7 +730,7 @@ Programs/_testembed: Programs/_testembed
 Programs/_freeze_importlib.o: Programs/_freeze_importlib.c Makefile
 
 Programs/_freeze_importlib: Programs/_freeze_importlib.o $(LIBRARY_OBJS_OMIT_FROZEN)
-	$(LINKCC) $(PY_CORE_LDFLAGS) -o $@ Programs/_freeze_importlib.o $(LIBRARY_OBJS_OMIT_FROZEN) $(LIBS) $(MODLIBS) $(SYSLIBS)
+	$(LINKCC) -fno-lto $(filter-out -flto% -ffat-lto-objects,$(PY_CORE_LDFLAGS)) -o $@ Programs/_freeze_importlib.o $(LIBRARY_OBJS_OMIT_FROZEN) $(LIBS) $(MODLIBS) $(SYSLIBS)
 
 .PHONY: regen-importlib
 regen-importlib: Programs/_freeze_importlib
