--- a/Lib/test/libregrtest/main.py
+++ b/Lib/test/libregrtest/main.py
@@ -607,6 +607,8 @@ class Regrtest:
             self.tmp_dir = self.ns.tempdir
 
         if not self.tmp_dir:
+            self.tmp_dir = tempfile.gettempdir()
+        if not self.tmp_dir:
             # When tests are run from the Python build directory, it is best practice
             # to keep the test files in a subfolder.  This eases the cleanup of leftover
             # files using the "make distclean" command.
--- a/Lib/test/support/socket_helper.py
+++ b/Lib/test/support/socket_helper.py
@@ -136,7 +136,7 @@ def _is_ipv6_enabled():
                 sock.close()
     return False
 
-IPV6_ENABLED = _is_ipv6_enabled()
+IPV6_ENABLED = False
 
 
 _bind_nix_socket_error = None
--- a/Lib/test/test__xxsubinterpreters.py
+++ b/Lib/test/test__xxsubinterpreters.py
@@ -489,7 +489,6 @@ class IsRunningTests(TestBase):
         main = interpreters.get_main()
         self.assertTrue(interpreters.is_running(main))
 
-    @unittest.skip('Fails on FreeBSD')
     def test_subinterpreter(self):
         interp = interpreters.create()
         self.assertFalse(interpreters.is_running(interp))
--- a/Lib/test/test_concurrent_futures.py
+++ b/Lib/test/test_concurrent_futures.py
@@ -377,10 +377,6 @@ class ExecutorShutdownTest:
         self.assertGreater(len(others), 0)
 
     def test_hang_gh83386(self):
-        """shutdown(wait=False) doesn't hang at exit with running futures.
-
-        See https://github.com/python/cpython/issues/83386.
-        """
         if self.executor_type == futures.ProcessPoolExecutor:
             raise unittest.SkipTest(
                 "Hangs, see https://github.com/python/cpython/issues/83386")
@@ -399,11 +395,6 @@ class ExecutorShutdownTest:
         self.assertEqual(out.strip(), b"apple")
 
     def test_hang_gh94440(self):
-        """shutdown(wait=True) doesn't hang when a future was submitted and
-        quickly canceled right before shutdown.
-
-        See https://github.com/python/cpython/issues/94440.
-        """
         if not hasattr(signal, 'alarm'):
             raise unittest.SkipTest(
                 "Tested platform does not support the alarm signal")
--- a/Lib/test/test_exceptions.py
+++ b/Lib/test/test_exceptions.py
@@ -1065,7 +1065,6 @@ class ExceptionTests(unittest.TestCase):
         self.assertIsInstance(exc.__context__, ValueError)
         self.assertIs(exc.__context__.__context__, exc.__context__)
 
-    @unittest.skip("See issue 44895")
     def test_no_hang_on_context_chain_cycle2(self):
         # See issue 25782. Cycle at head of context chain.
 
--- a/Lib/test/test_importlib/test_resource.py
+++ b/Lib/test/test_importlib/test_resource.py
@@ -190,7 +190,6 @@ class DeletingZipsTest(unittest.TestCase
         self.zip_path.unlink()
         del c
 
-    @unittest.skip("Desired but not supported.")
     def test_path_does_not_keep_open(self):
         c = resources.path('ziptestdata', 'binary.file')
         self.zip_path.unlink()
--- a/Lib/test/test_interpreters.py
+++ b/Lib/test/test_interpreters.py
@@ -244,7 +244,6 @@ class TestInterpreterIsRunning(TestBase)
         main = interpreters.get_main()
         self.assertTrue(main.is_running())
 
-    @unittest.skip('Fails on FreeBSD')
     def test_subinterpreter(self):
         interp = interpreters.create()
         self.assertFalse(interp.is_running())
@@ -372,7 +371,6 @@ class TestInterpreterClose(TestBase):
         t.start()
         t.join()
 
-    @unittest.skip('Fails on FreeBSD')
     def test_still_running(self):
         main, = interpreters.list_all()
         interp = interpreters.create()
@@ -430,7 +428,6 @@ class TestInterpreterRun(TestBase):
             content = file.read()
             self.assertEqual(content, expected)
 
-    @unittest.skip('Fails on FreeBSD')
     def test_already_running(self):
         interp = interpreters.create()
         with _running(interp):
--- a/Lib/test/test_nntplib.py
+++ b/Lib/test/test_nntplib.py
@@ -141,8 +141,6 @@ class NetworkedNNTPTestsMixin:
         self.assertLessEqual(art_num, last)
         self._check_art_dict(art_dict)
 
-    @unittest.skipIf(True, 'temporarily skipped until a permanent solution'
-                           ' is found for issue #28971')
     def test_over(self):
         resp, count, first, last, name = self.server.group(self.GROUP_NAME)
         start = last - 10
@@ -346,7 +344,7 @@ class NetworkedNNTPTests(NetworkedNNTPTe
         if cls.server is not None:
             cls.server.quit()
 
-@unittest.skipUnless(ssl, 'requires SSL support')
+@unittest.skip('skip tests for aioe.org')
 class NetworkedNNTP_SSLTests(NetworkedNNTPTests):
 
     # Technical limits for this public NNTP server (see http://www.aioe.org):
--- a/Lib/test/test_os.py
+++ b/Lib/test/test_os.py
@@ -3232,7 +3232,6 @@ class SpawnTests(unittest.TestCase):
 
 # The introduction of this TestCase caused at least two different errors on
 # *nix buildbots. Temporarily skip this to let the buildbots move along.
-@unittest.skip("Skip due to platform/environment differences on *NIX buildbots")
 @unittest.skipUnless(hasattr(os, 'getlogin'), "test needs os.getlogin")
 class LoginTests(unittest.TestCase):
     def test_getlogin(self):
--- a/Lib/test/test_pydoc.py
+++ b/Lib/test/test_pydoc.py
@@ -1009,7 +1009,6 @@ class PydocImportTest(PydocBaseTest):
             finally:
                 sys.path[:] = saved_paths
 
-    @unittest.skip('causes undesirable side-effects (#20128)')
     def test_modules(self):
         # See Helper.listmodules().
         num_header_lines = 2
@@ -1025,7 +1024,6 @@ class PydocImportTest(PydocBaseTest):
 
         self.assertGreaterEqual(num_lines, expected)
 
-    @unittest.skip('causes undesirable side-effects (#20128)')
     def test_modules_search(self):
         # See Helper.listmodules().
         expected = 'pydoc - '
@@ -1038,7 +1036,6 @@ class PydocImportTest(PydocBaseTest):
 
         self.assertIn(expected, result)
 
-    @unittest.skip('some buildbots are not cooperating (#20128)')
     def test_modules_search_builtin(self):
         expected = 'gc - '
 
--- a/Lib/test/test_socket.py
+++ b/Lib/test/test_socket.py
@@ -64,6 +64,7 @@ def get_cid():
 
 def _have_socket_can():
     """Check whether CAN sockets are supported on this host."""
+    return False
     try:
         s = socket.socket(socket.PF_CAN, socket.SOCK_RAW, socket.CAN_RAW)
     except (AttributeError, OSError):
@@ -74,6 +75,7 @@ def _have_socket_can():
 
 def _have_socket_can_isotp():
     """Check whether CAN ISOTP sockets are supported on this host."""
+    return False
     try:
         s = socket.socket(socket.PF_CAN, socket.SOCK_DGRAM, socket.CAN_ISOTP)
     except (AttributeError, OSError):
@@ -84,6 +86,7 @@ def _have_socket_can_isotp():
 
 def _have_socket_can_j1939():
     """Check whether CAN J1939 sockets are supported on this host."""
+    return False
     try:
         s = socket.socket(socket.PF_CAN, socket.SOCK_DGRAM, socket.CAN_J1939)
     except (AttributeError, OSError):
@@ -94,6 +97,7 @@ def _have_socket_can_j1939():
 
 def _have_socket_rds():
     """Check whether RDS sockets are supported on this host."""
+    return False
     try:
         s = socket.socket(socket.PF_RDS, socket.SOCK_SEQPACKET, 0)
     except (AttributeError, OSError):
@@ -114,6 +118,7 @@ def _have_socket_alg():
 
 def _have_socket_qipcrtr():
     """Check whether AF_QIPCRTR sockets are supported on this host."""
+    return False
     try:
         s = socket.socket(socket.AF_QIPCRTR, socket.SOCK_DGRAM, 0)
     except (AttributeError, OSError):
@@ -124,12 +129,14 @@ def _have_socket_qipcrtr():
 
 def _have_socket_vsock():
     """Check whether AF_VSOCK sockets are supported on this host."""
+    return False
     ret = get_cid() is not None
     return ret
 
 
 def _have_socket_bluetooth():
     """Check whether AF_BLUETOOTH sockets are supported on this host."""
+    return False
     try:
         # RFCOMM is supported by all platforms with bluetooth support. Windows
         # does not support omitting the protocol.
@@ -1030,6 +1037,7 @@ class GeneralModuleTests(unittest.TestCa
             with self.assertRaises(OSError, msg=explanation):
                 socket.gethostbyaddr(addr)
 
+    @unittest.skip('avoid sethostname()')
     @unittest.skipUnless(hasattr(socket, 'sethostname'), "test needs socket.sethostname()")
     @unittest.skipUnless(hasattr(socket, 'gethostname'), "test needs socket.gethostname()")
     def test_sethostname(self):
--- a/Lib/test/test_support.py
+++ b/Lib/test/test_support.py
@@ -65,7 +65,6 @@ class TestSupport(unittest.TestCase):
                         self.test_get_attribute)
         self.assertRaises(unittest.SkipTest, support.get_attribute, self, "foo")
 
-    @unittest.skip("failing buildbots")
     def test_get_original_stdout(self):
         self.assertEqual(support.get_original_stdout(), sys.stdout)
 
--- a/Lib/test/test_time.py
+++ b/Lib/test/test_time.py
@@ -131,6 +131,7 @@ class TimeTestCase(unittest.TestCase):
         self.assertGreater(res, 0.0)
         self.assertLessEqual(res, 1.0)
 
+    @unittest.skip("avoid clock_settime()")
     @unittest.skipUnless(hasattr(time, 'clock_settime'),
                          'need time.clock_settime()')
     def test_clock_settime(self):
@@ -432,10 +433,6 @@ class TimeTestCase(unittest.TestCase):
             else:
                 self.assertEqual(time.mktime(tt), t)
 
-    # Issue #13309: passing extreme values to mktime() or localtime()
-    # borks the glibc's internal timezone data.
-    @unittest.skipUnless(platform.libc_ver()[0] != 'glibc',
-                         "disabled because of a bug in glibc. Issue #13309")
     def test_mktime_error(self):
         # It may not be possible to reliably make mktime return an error
         # on all platforms.  This will make sure that no other exception
@@ -508,6 +505,7 @@ class TimeTestCase(unittest.TestCase):
         self.assertTrue(info.monotonic)
         self.assertFalse(info.adjustable)
 
+    @unittest.skip("avoid clock_settime()")
     @unittest.skipUnless(hasattr(time, 'clock_settime'),
                          'need time.clock_settime')
     def test_monotonic_settime(self):
