--- a/Makefile.pre.in
+++ b/Makefile.pre.in
@@ -520,7 +520,13 @@ build_all_generate_profile:
 
 run_profile_task:
 	@ # FIXME: can't run for a cross build
-	$(LLVM_PROF_FILE) $(RUNSHARED) ./$(BUILDPYTHON) $(PROFILE_TASK) || true
+	$(LLVM_PROF_FILE) $(RUNSHARED) ./$(BUILDPYTHON) $(PROFILE_TASK) || true ; \
+	[ "$${PGO_EXTRA_PASS}" = 1 ] || exit 0 ; \
+	for i in 1 2 ; do \
+	  PYTHONPATH='$(PGO_PYPERF_MODULES)' \
+	  $(LLVM_PROF_FILE) $(RUNSHARED) ./$(BUILDPYTHON) -m pyperformance run --debug-single-value || true ; \
+	  unset CONTAINER_PYTHON_COMPAT ; \
+	done
 
 build_all_merge_profile:
 	$(LLVM_PROF_MERGER)
--- a/configure.ac
+++ b/configure.ac
@@ -1341,7 +1341,7 @@ AC_ARG_VAR(PROFILE_TASK, Python args for
 AC_MSG_CHECKING(PROFILE_TASK)
 if test -z "$PROFILE_TASK"
 then
-	PROFILE_TASK='-m test --pgo --timeout=$(TESTTIMEOUT)'
+	PROFILE_TASK='-m test --pgo-extended --timeout=$(TESTTIMEOUT)'
 fi
 AC_MSG_RESULT($PROFILE_TASK)
 
