--- a/Python/initconfig.c
+++ b/Python/initconfig.c
@@ -790,6 +790,9 @@ config_init_defaults(PyConfig *config)
 {
     _PyConfig_InitCompatConfig(config);
 
+    int compat_behavior = 0;
+    _Py_get_env_flag(1, &compat_behavior, "CONTAINER_PYTHON_COMPAT");
+
     config->isolated = 0;
     config->use_environment = 1;
     config->site_import = 1;
@@ -798,11 +801,11 @@ config_init_defaults(PyConfig *config)
     config->interactive = 0;
     config->optimization_level = 0;
     config->parser_debug= 0;
-    config->write_bytecode = 1;
+    config->write_bytecode = (compat_behavior) ? 1 : 0;
     config->verbose = 0;
     config->quiet = 0;
     config->user_site_directory = 1;
-    config->buffered_stdio = 1;
+    config->buffered_stdio = (compat_behavior) ? 1 : 0;
     config->pathconfig_warnings = 1;
 #ifdef MS_WINDOWS
     config->legacy_windows_stdio = 0;
