--- a/Lib/test/libregrtest/cmdline.py
+++ b/Lib/test/libregrtest/cmdline.py
@@ -164,6 +164,8 @@ class Namespace(argparse.Namespace):
         self.match_tests: TestFilter = []
         self.pgo = False
         self.pgo_extended = False
+        self.lean_pgo = False
+        self.lean_pgo_extended = False
         self.worker_json = None
         self.start = None
         self.timeout = None
@@ -334,6 +336,8 @@ def _create_parser():
                        help='enable Profile Guided Optimization (PGO) training')
     group.add_argument('--pgo-extended', action='store_true',
                        help='enable extended PGO training (slower training)')
+    group.add_argument('--lean-pgo', action='store_true', help='internal option')
+    group.add_argument('--lean-pgo-extended', action='store_true', help='internal option')
     group.add_argument('--fail-env-changed', action='store_true',
                        help='if a test file alters the environment, mark '
                             'the test as failed')
@@ -465,6 +469,8 @@ def _parse_args(args, **kwargs):
         parser.error("--pgo/-v don't go together!")
     if ns.pgo_extended:
         ns.pgo = True  # pgo_extended implies pgo
+    if ns.lean_pgo_extended:
+        ns.lean_pgo = True
 
     if ns.nowindows:
         print("Warning: the --nowindows (-n) option is deprecated. "
--- a/Lib/test/libregrtest/main.py
+++ b/Lib/test/libregrtest/main.py
@@ -55,6 +55,8 @@ class Regrtest:
         self.quiet: bool = ns.quiet
         self.pgo: bool = ns.pgo
         self.pgo_extended: bool = ns.pgo_extended
+        self.lean_pgo: bool = ns.lean_pgo
+        self.lean_pgo_extended: bool = ns.lean_pgo_extended
 
         # Test results
         self.results: TestResults = TestResults()
@@ -181,6 +183,8 @@ class Regrtest:
         if self.pgo:
             # add default PGO tests if no tests are specified
             setup_pgo_tests(self.cmdline_args, self.pgo_extended)
+        if self.lean_pgo:
+            setup_pgo_tests(self.cmdline_args, self.lean_pgo_extended)
 
         exclude_tests = set()
         if self.exclude:
--- a/Makefile.pre.in
+++ b/Makefile.pre.in
@@ -641,7 +641,7 @@ build_all_generate_profile:
 
 run_profile_task:
 	@ # FIXME: can't run for a cross build
-	$(LLVM_PROF_FILE) $(RUNSHARED) ./$(BUILDPYTHON) $(PROFILE_TASK) || true
+	$(LLVM_PROF_FILE) $(RUNSHARED) $(DEB_SRC_TOPDIR)/debian/bench.sh ./$(BUILDPYTHON)
 
 build_all_merge_profile:
 	$(LLVM_PROF_MERGER)
